### PR TITLE
Add Metadata to Mesh

### PR DESCRIPTION
**Pull request by runette**
Working on making the Python interface write enabled and making the PLY driver write-enabled, I have realised that there is a need for metadata on the mesh. The primary use case for this is to retain format related information (in the PLY case e.g. whether the file is text or binary) to allow round trip consistency.

My preferred way to do this is to add the same metadata capabilities to the mesh object as is currently found on the DatasetGroup object (rather than try to proliferate driver specific functionality).

I am proposing that we basically copy the metadata related methods and functionality from the DatasetGroup object to the Mesh object to provide consistency.

This is a NON BREAKING change to the API - since it does not change any current functionality and does not require that the driver uses the metadata in any way. It will, however, allow more flexibility for future drivers etc.

### CODE REVIEW

**Review by PeterPetrik — APPROVED** [review]

File: mdal/mdal_data_model.hpp [hunk 1]
```diff
@@ -152,7 +152,7 @@ namespace MDAL
 
       std::string getMetadata( const std::string &key );
       void setMetadata( const std::string &key, const std::string &val );
-      void setMetadata( const Metadata &metadata );
+      void setMetadata( const Metadata &new_metadata );
```
Comment (PeterPetrik) [inline mdal/mdal_data_model.hpp:155]: why the change of the arg name?

File: mdal/mdal_data_model.hpp [hunk 2]
```diff
@@ -267,6 +267,12 @@ namespace MDAL
       std::string crs() const;
       size_t faceVerticesMaximumCount() const;
 
+      std::string getMetadata( const std::string &key );
```
Comment (PeterPetrik) [inline mdal/mdal_data_model.hpp:270]: we do not use getXYZ() syntax for getters, can you rename to `std::string metadata(...)` ?

File: mdal/mdal_data_model.hpp (hunk 1 above)
Comment (runette) [reply mdal/mdal_data_model.hpp:155]: because having the same name provokes a compiler warning in windows since the variable hides a class member.

Not a big problem - but it does make sense to avoid the compiler warnings.

File: mdal/mdal_data_model.hpp (hunk 2 above)
Comment (runette) [reply mdal/mdal_data_model.hpp:270]: I actually copied the structure and names exactly from DatasetGroup - including the getter - for complete consistency.

I cannot rename to `metadata` - because there is already a class member Mesh::metadata (and DatasetGroup::metadata since the former is an exact copy of the latter in structure).

The `getMetadata()` method does add the ability to get metadata by key as opposed to by index. In the DatasetGroup case this is used in one place  DAT driver.

I can rename the method anything you want (except metadata) but I do strongly suggest that we keep the signatures for Mesh metadata and DatasetGroup metadata consistent, otherwise we are just creating work and potential errors.

Comment (PeterPetrik) [reply mdal/mdal_data_model.hpp:270]: ok
